File: Data Structures/Data Structures/Sorts/Bubble/bubble.java
package Sorts.Bubble;

public class bubble {
    /**
     * Method that uses bubble sort to sort a generic array into ascending order. 
     *
     * @param arr   the input array to sort
     * @return      the sorted array
     */
    public static <T> T[] bubbleAsc(T[] arr) {
        // for each pass in the sort
        for (int i=0; i < arr.length; i++) {
            // for every unsorted element
            for (int j=0; j < arr.length - 1 - i; j++) {
                // swap them
                if (arr[j].toString().compareTo(arr[j+1].toString()) > 0) {
                    T temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                }
            }
        }
        return arr;
    }

    /**
     * Method that uses bubble sort to sort a generic array into descending order. 
     *
     * @param arr   the input array to sort
     * @return      the sorted array
     */
    public static <T> T[] bubbleDesc(T[] arr) {
        // for each pass in the sort
        for (int i=0; i < arr.length; i++) {
            // for every unsorted element
            for (int j=0; j < arr.length - 1 - i; j++) {
                // swap them
                if (arr[j].toString().compareTo(arr[j+1].toString()) < 0) {
                    T temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                }
            }
        }
        return arr;
    }

    public static void main(String[] args) {
        Integer[] o = {1, 4, 6, 5, 3, 2};
        o = bubbleAsc(o);
        for (int p: o) {
            System.out.print(p+" ");
        }
        System.out.print('\n');
        o = bubbleDesc(o);
        for (int p: o) {
            System.out.print(p+" ");
        }
    }
}

File: Data Structures/Data Structures/Sorts/Bubble/bubble.py
# function to sort in ascending order
def bubbleAscending(arr):
    # for each pass in the bubble sort
    for i in range(0, len(arr)):
        # for each unsorted item in the list
        for j in range(0, len(arr)-1-i):
            # if the current item is smaller than the next item, swap them
            if arr[j] > arr[j+1]:
                temp = arr[j]
                arr[j] = arr[j+1]
                arr[j+1] = temp
    # returns a sorted array
    return arr

# function to sort in ascending order
def bubbleDescending(arr):
    for i in range(0, len(arr)):
        # for each unsorted item in the list
        for j in range(0, len(arr)-1-i):
            # if the current item is larger than the next item, swap them
            if arr[j] < arr[j+1]:
                temp = arr[j]
                arr[j] = arr[j+1]
                arr[j+1] = temp
    # returns a sorted array
    return arr

# tests
print(bubbleAscending([1, 6, 7, 3, 4, 5, 2]))
print(bubbleDescending([1, 6, 7, 3, 4, 5, 2]))

# outputs
'''
[1, 2, 3, 4, 5, 6, 7]
[7, 6, 5, 4, 3, 2, 1]
'''